### PR TITLE
make d3-tip work with d3js v4

### PR DESCRIPTION
**Pull request by besson-genostar**
Another nice thing to do would be making d3-tip a d3js plugin/module

### CODE REVIEW

**Comment by vdh** [comment]
👍  This looks like it could be a winner for #156, the other two v4 forks stripped out too much code

File: index.js [hunk 1]
```diff
@@ -51,29 +51,30 @@
           i       = directions.length,
           coords,
           scrollTop  = document.documentElement.scrollTop || document.body.scrollTop,
-          scrollLeft = document.documentElement.scrollLeft || document.body.scrollLeft
+          scrollLeft = document.documentElement.scrollLeft || document.body.scrollLeft;
 
       nodel.html(content)
-        .style({ opacity: 1, 'pointer-events': 'all' })
+        .style('opacity', 1).style('pointer-events', 'all');
```
Comment (Caged) [inline index.js]: Did the previous hash-style syntax stop working?

File: index.js [hunk 2]
```diff
@@ -51,29 +51,30 @@
...
-        top: (coords.top +  poffset[0]) + scrollTop + 'px',
-        left: (coords.left + poffset[1]) + scrollLeft + 'px'
-      })
+      while(i--) {
+      	nodel.classed(directions[i], false);
+      }
+      coords = direction_callbacks.get(dir).apply(this);
+      nodel.classed(dir, true)
+      	.style('top', (coords.top +  poffset[0]) + scrollTop + 'px')
+      	.style('left', (coords.left + poffset[1]) + scrollLeft + 'px');;
```
Comment (Caged) [inline index.js]: ^ Same concern as above regarding hash-style syntax working.

File: index.js [hunk 3]
```diff
@@ -116,35 +117,41 @@
...
-      direction = v == null ? v : d3.functor(v)
+      if (!arguments.length) {
+      	return direction;
+      }
+      direction = v == null ? v : functor(v);
 
-      return tip
-    }
+      return tip;
+    };
```
Comment (Caged) [inline index.js]: This seems to be a personal formatting preference.  Could you collapse this like the previous code?

File: index.js [hunk 4]
```diff
@@ -116,35 +117,41 @@
...
-      offset = v == null ? v : d3.functor(v)
+      if (!arguments.length) {
+      	return offset;
+      }
+      offset = v == null ? v : functor(v);
 
-      return tip
-    }
+      return tip;
+    };
```
Comment (Caged) [inline index.js]: Same, this looks like mostly formatting changes.  Could you collapse this like the previous version?

**Comment by Caged** [comment]
Thanks for the PR @besson-genostar!  

I would prefer to keep this PR to the minimal changes required to get this working with d3 v4.  It looks like this contains some general formatting changes and the addition of semicolons which makes the PR larger than necessary.  Would you mind updating this PR or submitting a new one that doesn't add semicolons and change the format where it's unnecessary?

File: index.js (hunk 1 above)
Comment (besson-genostar) [inline index.js]: Yes. It has been moved to a plugin which is not included by default

From https://github.com/d3/d3-selection/blob/master/README.md

> For the sake of parsimony, the multi-value methods—where you pass an object to set multiple attributes, styles or properties simultaneously—have been extracted to d3-selection-multi and are no longer part of the default bundle.

File: index.js (hunk 3 above)
Comment (besson-genostar) [inline index.js]: Yeah sorry, this was mostly automatic. I'll switch back to previous formatting.

**Comment by besson-genostar** [comment]
Is this OK this way?

**Comment by Caged** [comment]
@besson-genostar thanks for the quick updates!  It looks like you missed changing the hash-style syntax.  Would you mind making that change?

**Comment by besson-genostar** [comment]
I answered your comment about this: hash-style syntax is now part of a module which is not included by default in the d3js v4 bundle, thus I think it's best to keep to minimal requirements. Do you agree with this? The changes look quite simple to me and do not add much complexity.

**Comment by Caged** [comment]
@besson-genostar Apologies, your comment got collapsed and I missed it.  That works for me 👍

**Pull request merged**